ExampleMod/Content/TileEntities/BasicTileEntity.cs

```diff
@@ -0,0 +1,243 @@
+				Point16 topLeft = TileObjectData.TopLeft(i, j);
+				if (!Main.sectionManager.TileLoaded(topLeft.X, topLeft.Y)) {
+					return $"{name}: World section not loaded yet";
+				}
+				return $"{name}: No TileEntity found at coordinate";
+			}
+		}
+
+		public override void MouseOver(int i, int j) {
+			if (TileEntity.TryGet(i, j, out BasicTileEntity tileEntity)) {
+				Player player = Main.LocalPlayer;
+				player.noThrow = 2;
+				player.cursorItemIconEnabled = true;
+				player.cursorItemIconID = -1;
+				player.cursorItemIconText = StatusText.Format(tileEntity.WaterFillPercentage);
+			}
+		}
+
+		public override bool RightClick(int i, int j) {
+			if (!TileEntity.TryGet(i, j, out BasicTileEntity tileEntity)) {
+				return true;
+			}
+			if (tileEntity.WaterFillPercentage == 100) {
+				/* While it is tempting to spawn the item like this, this will not work correctly in multiplayer. See below for the correct approach.
+				Main.LocalPlayer.QuickSpawnItem(Main.LocalPlayer.GetSource_TileInteraction(i, j), ItemID.WaterBucket);
+				tileEntity.WaterFillLevel = 0;
+				*/
+				SoundEngine.PlaySound(SoundID.Drown);
+
+				// This example follows the convention of similar vanilla tile entitites that drop items: When right clicked, the tile will be "mined" and the item is released by the KillTile code running on the server, which will automatically spawn the item and adjust the tile entity data on the server preventing various potential network syncing issues.
+				WorldGen.KillTile(i, j, fail: true);
+				if (Main.netMode == NetmodeID.MultiplayerClient) {
+					// This is sending the KillTile manipulation for the (i, j) coordinates. number4 being 1f means KillTile should fail. This is what will cause KillTile to run on the server, ultimately spawning the Water Bucket item and setting WaterFillLevel back to 0. 
+					NetMessage.SendData(MessageID.TileManipulation, number: 0, number2: i, number3: j, number4: 1f);
+				}
+				// You can use a different approach for the item dropping, if you prefer, but care must be taken to ensure that multiple users interacting with the tile do not result in duplicate item spawns due to network latency or desync. This means using a custom ModPacket to inform the server to do something rather than running the code on the client.
+			}
+			else {
+				// If not full, show a chat message showing the fill percentage.
+				Main.NewText(StatusText.Format(tileEntity.WaterFillPercentage));
+			}
+
+			return true;
+		}
+
+		public override void KillTile(int i, int j, ref bool fail, ref bool effectOnly, ref bool noItem) {
+			// Some tile entities that hold items, like Weapon Rack and Item Frame, release their item when hit with a pickaxe. When doing this they will not be destroyed on the first hit no matter how strong the pickaxe is.
+			// Others like Hat Rack and Mannequin do not release the item and prevent the tile from being mined at all.
+
```

Comment:
A line here noting that the `TIleManipulation` packet sent from `RightClick` will have `fail = true` so we just run this code to re-check if there's an item to drop.